mongosql/src/algebrizer/definitions.rs

```diff
@@ -475,14 +478,105 @@ impl<'a> Algebrizer<'a> {
         }
 
         // Build the Project Stage using the source and built expression.
-        let stage = mir::Stage::Project(mir::Project {
+        let project_stage = mir::Stage::Project(mir::Project {
             source: Box::new(source),
-            expression,
+            expression: expression.clone(),
             is_add_fields,
             cache: SchemaCache::new(),
         });
-        stage.schema(&self.schema_inference_state())?;
-        Ok(stage)
+        project_stage.schema(&self.schema_inference_state())?;
+        Ok(if is_distinct {
+            let group_stage = Self::build_distinct_values_group_and_project_stage(
+                project_stage,
+                self.scope_level,
+                datasources,
+            );
+            group_stage.schema(&self.schema_inference_state())?;
+            group_stage
+        } else {
+            project_stage
+        })
+    }
+
+    pub fn build_distinct_values_group_and_project_stage(
+        source_stage: mir::Stage,
+        scope_level: u16,
+        datasources: BTreeSet<Key>,
+    ) -> mir::Stage {
+        let mut group_keys = Vec::new();
+        let mut bottom: Option<AliasedExpr> = None;
+        let mut datasource_aliases = Vec::new();
+
+        // Identify data sources and build group keys for deduplication
+        for key in datasources.iter() {
+            if key.scope == scope_level {
```

Comment:
As opposed to the `*` case below, in this case we know that `datasources` is exactly the set of datasource bindings being output from the `source` so we actually can safely ignore this `scope_level` check!